File: examples/excel-add-image.ts
/**
 * 例子（Epic-13）：从零构造一份 xlsx，在第 0 张 sheet 的 (B2, F10) 范围嵌一张 PNG。
 *
 * 跑法：
 *   bun run examples/excel-add-image.ts <output.xlsx>
 */

import { SpreadsheetDocument, createImageTwoCellAnchorForExcel } from "../src/excel/index.js";

// 1×1 透明 PNG。
const TINY_PNG = new Uint8Array([
  0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x48, 0x44, 0x52,
  0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1f, 0x15, 0xc4,
  0x89, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x44, 0x41, 0x54, 0x78, 0x9c, 0x62, 0x00, 0x01, 0x00, 0x00,
  0x05, 0x00, 0x01, 0x0d, 0x0a, 0x2d, 0xb4, 0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4e, 0x44, 0xae,
  0x42, 0x60, 0x82,
]);

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: excel-add-image <output.xlsx>\n");
    process.exit(2);
  }

  const doc = SpreadsheetDocument.create();
  const { relId, drawingPart } = doc.addImagePart(0, TINY_PNG);

  // (col 1, row 1) = B2；(col 5, row 9) = F10。0-based 索引。
  const anchor = createImageTwoCellAnchorForExcel(
    relId,
    { col: 1, row: 1 },
    { col: 5, row: 9 },
    { id: 1, name: "Logo" },
  );
  drawingPart.wsDr.appendChild(anchor);

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (PNG embedded on sheet1, B2:F10)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/ExcelAddImage.cs
// Replica of examples/excel-add-image.ts
// Creates an xlsx with a 1x1 PNG anchored to B2:F10 using a TwoCellAnchor.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Spreadsheet;
using A = DocumentFormat.OpenXml.Drawing;
using XDR = DocumentFormat.OpenXml.Drawing.Spreadsheet;

namespace CrossSdkVerify.Replicas;

public static class ExcelAddImage
{
    private static readonly byte[] TinyPng =
    [
        0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x48, 0x44, 0x52,
        0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1f, 0x15, 0xc4,
        0x89, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x44, 0x41, 0x54, 0x78, 0x9c, 0x62, 0x00, 0x01, 0x00, 0x00,
        0x05, 0x00, 0x01, 0x0d, 0x0a, 0x2d, 0xb4, 0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4e, 0x44, 0xae,
        0x42, 0x60, 0x82,
    ];

    public static void Run(string outputPath)
    {
        using var doc = SpreadsheetDocument.Create(outputPath,
            DocumentFormat.OpenXml.SpreadsheetDocumentType.Workbook);

        var wbPart = doc.AddWorkbookPart();
        wbPart.Workbook = new Workbook();

        var wsPart = wbPart.AddNewPart<WorksheetPart>();
        var ws = new Worksheet(new SheetData());

        var sheets = wbPart.Workbook.AppendChild(new Sheets());
        sheets.Append(new Sheet
        {
            Id = wbPart.GetIdOfPart(wsPart),
            SheetId = 1,
            Name = "Sheet1"
        });

        // Add drawing part
        var drawingPart = wsPart.AddNewPart<DrawingsPart>();
        var drawingRelId = wsPart.GetIdOfPart(drawingPart);

        // Add image part to drawing part
        var imagePart = drawingPart.AddImagePart(ImagePartType.Png);
        using (var ms = new MemoryStream(TinyPng))
            imagePart.FeedData(ms);
        var imgRelId = drawingPart.GetIdOfPart(imagePart);

        // WorksheetDrawing
        drawingPart.WorksheetDrawing = new XDR.WorksheetDrawing(
            new XDR.TwoCellAnchor(
                new XDR.FromMarker(
                    new XDR.ColumnId("1"),         // col 1 = B (0-based)
                    new XDR.ColumnOffset("0"),
                    new XDR.RowId("1"),             // row 1 = 2 (0-based)
                    new XDR.RowOffset("0")),
                new XDR.ToMarker(
                    new XDR.ColumnId("5"),          // col 5 = F (0-based)
                    new XDR.ColumnOffset("0"),
                    new XDR.RowId("9"),             // row 9 = 10 (0-based)
                    new XDR.RowOffset("0")),
                new XDR.Picture(
                    new XDR.NonVisualPictureProperties(
                        new XDR.NonVisualDrawingProperties { Id = 1, Name = "Logo" },
                        new XDR.NonVisualPictureDrawingProperties()),
                    new XDR.BlipFill(
                        new A.Blip { Embed = imgRelId },
                        new A.Stretch(new A.FillRectangle())),
                    new XDR.ShapeProperties(
                        new A.Transform2D(
                            new A.Offset { X = 0, Y = 0 },
                            new A.Extents { Cx = 0, Cy = 0 }),
                        new A.PresetGeometry(new A.AdjustValueList())
                        { Preset = A.ShapeTypeValues.Rectangle })),
                new XDR.ClientData()
            )
        );

        // Link drawing to worksheet
        ws.Append(new Drawing { Id = drawingRelId });
        wsPart.Worksheet = ws;
        wbPart.Workbook.Save();
    }
}
